plutus-contract/src/Plutus/Contract/Marconi/Handler.hs

```diff
@@ -0,0 +1,176 @@
+import Control.Monad.Freer.Error (Error, runError, throwError)
+import Control.Monad.Freer.Extras (raiseMUnderN)
+import Control.Monad.Freer.Extras.Pagination (PageQuery, pageOf)
+import Control.Monad.Freer.State qualified as Eff (State, get, put, runState)
+import Control.Monad.Freer.TH (makeEffect)
+import Control.Monad.IO.Class (MonadIO (liftIO))
+import Data.Foldable (foldl')
+import Data.Maybe (fromJust)
+import Data.Set qualified as Set
+import Ledger (CardanoTx (CardanoTx))
+import Ledger.Address (CardanoAddress)
+import Ledger.Tx.CardanoAPI.Internal (fromCardanoTxId)
+import Marconi.ChainIndex.Indexers.Utxo (StorableEvent (UtxoEvent), StorableQuery (UtxoAddress), UtxoHandle, getInputs,
+                                         getUtxoResult, getUtxos, txId, txIx, urUtxo)
+import Marconi.Core.Storable (HasPoint, QueryInterval (QEverything), Queryable, State, StorableMonad, StorablePoint,
+                              StorableResult, insertMany, query)
+import Plutus.ChainIndex.Api (UtxosResponse (UtxosResponse))
+import Plutus.ChainIndex.ChainIndexError (ChainIndexError (..))
+import Plutus.ChainIndex.Compatibility (toCardanoBlockId)
+import Plutus.ChainIndex.Effects (ChainIndexControlEffect (AppendBlocks), ChainIndexQueryEffect (UtxoSetAtAddress))
+import Plutus.ChainIndex.Types (ChainSyncBlock (..), Point (Point, PointAtGenesis), Tip (TipAtGenesis), citxCardanoTx,
+                                tipAsPoint)
+import Plutus.V2.Ledger.Api (TxOutRef (TxOutRef))
+
+{- Developer note:
+   How to add new indexer to support new effect?
+       1. add the indexer MVar to `ChainIndexIndexersMvar` and the corresponding indexer to `ChainIndexIndexers`
+       2. edit `getChainIndexIndexers` and `putChainIndexIndexers` accordingly
+       3. generate `MarconiEffect` on the appropriate queries of `ChainIndexQueries` in `handleQuery`
+       4. Add the indexer update in the control operations
+       5. Add the `handleMarconiQuery` for the new effect in `handleChainIndexEffects`
+-}
+
+data ChainIndexIndexers -- We don't use `newtype` since other indexers will be needed
+   = ChainIndexIndexers
+       { _utxosIndexer :: State UtxoHandle
+       }
+
+makeLenses ''ChainIndexIndexers
+
+data ChainIndexIndexersMVar -- We don't use `newtype` since other indexers will be needed
+   = ChainIndexIndexersMVar
+       { _utxosIndexerMVar :: MVar (State UtxoHandle)
+       }
+
+makeLenses ''ChainIndexIndexersMVar
+
+getChainIndexIndexers :: ChainIndexIndexersMVar -> IO ChainIndexIndexers
+getChainIndexIndexers mvarCi =
+    ChainIndexIndexers <$> readMVar (mvarCi ^. utxosIndexerMVar)
+
+putChainIndexIndexers :: ChainIndexIndexers -> ChainIndexIndexersMVar -> IO ()
+putChainIndexIndexers ci mvarCi = do
+    putMVar (mvarCi ^. utxosIndexerMVar) (ci ^. utxosIndexer)
+
+
+
+data MarconiEffect handle r where
+  QueryIndexer :: StorableQuery handle -> MarconiEffect handle (StorableResult handle)
```

Comment:
The idea is that you may want to mock the `MarconiEffect` for a specific handle/indexer. On second thought, you're probably right as if we want to change it, we'll probably have to change the global handler for `ChainIndexQuery`.
It may help to slice the queries by indexer though.

Reply:
Your call :) Maybe a comment on that effect explaining the rationale would be enough for me